File: games/csharp/Pong.cs
using Tinkerforge;
using System.Threading;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.Serialization.Formatters.Binary;
using System.IO;

class PongSpeaker
{
	private bool okay = false;
	private BrickletPiezoSpeaker speaker = null;
	private Thread thread = null;

	public PongSpeaker(IPConnection ipcon)
	{
		if (Config.UID_PIEZO_SPEAKER_BRICKLET == null)
		{
			System.Console.WriteLine("Not Configured: Piezo Speaker");
			return;
		}

		speaker = new BrickletPiezoSpeaker(Config.UID_PIEZO_SPEAKER_BRICKLET, ipcon);

		try
		{
			string uid;
			string connectedUid;
			char position;
			byte[] hardwareVersion;
			byte[] firmwareVersion;
			int deviceIdentifier;

			speaker.GetIdentity(out uid, out connectedUid, out position,
			                    out hardwareVersion, out firmwareVersion,
			                    out deviceIdentifier);
			System.Console.WriteLine("Found: Piezo Speaker ({0})",
			                         Config.UID_PIEZO_SPEAKER_BRICKLET);
		}
		catch (TinkerforgeException)
		{
			System.Console.WriteLine("Not Found: Piezo Speaker ({0})",
			                         Config.UID_PIEZO_SPEAKER_BRICKLET);
			return;
		}

		okay = true;
	}

	public void Sirene(int freq)
	{
		if (!okay)
		{
			return;
		}

		for (int j = 0; j < 2; ++j)
		{
			for (int i = 0; i < 25; ++i)
			{
				speaker.Beep(10, freq + i * 20);
				Thread.Sleep(7);
			}

			for (int i = 0; i < 25; ++i)
			{
				speaker.Beep(10, freq + 24 * 20 - i * 20);
				Thread.Sleep(7);
			}
		}
	}

	public void BeepPaddleHit()
	{
		if (!okay)
		{
			return;
		}

		speaker.Beep(100, 500);
	}

	public void BeepSirene()
	{
		if (!okay)
		{
			return;
		}

		thread = new Thread(delegate() { Sirene(1000); });
		thread.IsBackground = true;
		thread.Start();
	}
}

class Pong
{
	private static int PADDLE_SIZE = 3;
	private static int CHUNK_SIZE = 16;

	private static byte[][] COLORS = new byte[][]
	{
		new byte[] {  0,   0,   0}, // off
//		new byte[] { 10,  10,  10}, // grey
		new byte[] {255,   0,   0}, // red
		new byte[] {255,  80,   0}, // orange
		new byte[] {255, 255,   0}, // yellow
		new byte[] {  0, 255,   0}, // green
		new byte[] {  0,   0, 255}, // blue
		new byte[] {255,   0, 150}, // violet
		new byte[] {255,   0,  40}, // purple
	};

	private static byte[][][] SCORE_FONT = new byte[][][]
	{
		new byte[][] { new byte[] {2, 2, 2},
		               new byte[] {2, 0, 2},
		               new byte[] {2, 0, 2},
		               new byte[] {2, 0, 2},
		               new byte[] {2, 2, 2}},
		new byte[][] { new byte[] {0, 2, 0},
		               new byte[] {0, 2, 0},
		               new byte[] {0, 2, 0},
		               new byte[] {0, 2, 0},
		               new byte[] {0, 2, 0}},
		new byte[][] { new byte[] {2, 2, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {2, 2, 2},
		               new byte[] {2, 0, 0},
		               new byte[] {2, 2, 2}},
		new byte[][] { new byte[] {2, 2, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {2, 2, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {2, 2, 2}},
		new byte[][] { new byte[] {2, 0, 2},
		               new byte[] {2, 0, 2},
		               new byte[] {2, 2, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {0, 0, 2}},
		new byte[][] { new byte[] {2, 2, 2},
		               new byte[] {2, 0, 0},
		               new byte[] {2, 2, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {2, 2, 2}},
		new byte[][] { new byte[] {2, 2, 2},
		               new byte[] {2, 0, 0},
		               new byte[] {2, 2, 2},
		               new byte[] {2, 0, 2},
		               new byte[] {2, 2, 2}},
		new byte[][] { new byte[] {2, 2, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {0, 0, 2}},
		new byte[][] { new byte[] {2, 2, 2},
		               new byte[] {2, 0, 2},
		               new byte[] {2, 2, 2},
		               new byte[] {2, 0, 2},
		               new byte[] {2, 2, 2}},
		new byte[][] { new byte[] {2, 2, 2},
		               new byte[] {2, 0, 2},
		               new byte[] {2, 2, 2},
		               new byte[] {0, 0, 2},
		               new byte[] {0, 0, 2}}
	};

	private BrickletLEDStrip ledStrip = null;
	private bool okay = false;
	private bool loop = true;
	private Timer timer = null;
	private byte[][] playfield = new byte[Config.LED_ROWS][];
	private int[] score = new int[] {0, 0};
	private int[] paddlePositionX = new int[] {4, 15};
	private int[] paddlePositionY = new int[] {3, 3};
	private double[] ballPosition = new double[] {10, 5};
	private double[] ballDirection = new double[] {0.1, 0.2};
	private KeyPress keyPress = null;
	private PongSpeaker speaker = null;

	public Pong(IPConnection ipcon)
	{
		for (int row = 0; row < Config.LED_ROWS; ++row)
		{
			playfield[row] = new byte[Config.LED_COLS];
		}

		// Call a getter to check that the Bricklet is avialable
		ledStrip = new BrickletLEDStrip(Config.UID_LED_STRIP_BRICKLET, ipcon);

		try
		{
			ledStrip.GetFrameDuration();
			System.Console.WriteLine("Found: LED Strip ({0})", Config.UID_LED_STRIP_BRICKLET);
		}
		catch (TinkerforgeException)
		{
			System.Console.WriteLine("Not Found: LED Strip ({0})", Config.UID_LED_STRIP_BRICKLET);
			return;
		}

		keyPress = new KeyPress(ipcon);
		speaker = new PongSpeaker(ipcon);
		okay = true;

		ledStrip.SetFrameDuration(40);
		ledStrip.FrameRendered += FrameRenderedCB;

		InitGame();
	}

	public bool IsOkay()
	{
		return okay;
	}

	private void InitGame()
	{
		NewBall();

		paddlePositionY[0] = 3;
		paddlePositionY[1] = 3;
		score[0] = 0;
		score[1] = 0;
	}

	private void FrameRenderedCB(BrickletLEDStrip sender, int length)
	{
		WritePlayfield();
	}

	private static T DeepCopy<T>(T obj)
	{
		MemoryStream stream = new MemoryStream();
		BinaryFormatter formatter = new BinaryFormatter();

		formatter.Serialize(stream, obj);
		stream.Position = 0;

		return (T)formatter.Deserialize(stream);
	}

	private void WritePlayfield()
	{
		if (!okay)
		{
			return;
		}

		byte[][] field = DeepCopy(playfield);

		AddScoreToPlayfield(field);
		AddPaddlesToPlayfield(field);
		AddBallToPlayfield(field);

		// Reorder LED data into R, G and B channel
		byte[] r = new byte[Config.LED_ROWS * Config.LED_COLS];
		byte[] g = new byte[Config.LED_ROWS * Config.LED_COLS];
		byte[] b = new byte[Config.LED_ROWS * Config.LED_COLS];

		for (int row = 0, i = 0; row < Config.LED_ROWS; ++row)
		{
			int colBegin;
			int colEnd;
			int colStep;

			if (row % 2 == 0)
			{
				colBegin = Config.LED_COLS - 1;
				colEnd = -1;
				colStep = -1;
			}
			else
			{
				colBegin = 0;
				colEnd = Config.LED_COLS;
				colStep = 1;
			}

			for (int col = colBegin; col != colEnd; col += colStep, ++i)
			{
				r[i] = COLORS[field[row][col]][Config.R_INDEX];
				g[i] = COLORS[field[row][col]][Config.G_INDEX];
				b[i] = COLORS[field[row][col]][Config.B_INDEX];
			}
		}

		// Make chunks of size 16
		byte[] rChunk = new byte[CHUNK_SIZE];
		byte[] gChunk = new byte[CHUNK_SIZE];
		byte[] bChunk = new byte[CHUNK_SIZE];

		for (int i = 0; i < Config.LED_ROWS * Config.LED_COLS; i += CHUNK_SIZE)
		{
			byte k;

			for (k = 0; k < CHUNK_SIZE && i + k < Config.LED_ROWS * Config.LED_COLS; ++k)
			{
				rChunk[k] = r[i + k];
				gChunk[k] = g[i + k];
				bChunk[k] = b[i + k];
			}

			try
			{
				ledStrip.SetRGBValues(i, k, rChunk, gChunk, bChunk);
			}
			catch (TinkerforgeException)
			{
				break;
			}
		}
	}

	private void AddScoreToPlayfield(byte[][] field)
	{
		for (int row = 0; row < 3; ++row)
		{
			for (int col = 0; col < 5; ++col)
			{
				field[row][col+1] = SCORE_FONT[score[0]][col][row];
				field[row+17][col+1] = SCORE_FONT[score[1]][col][row];
			}
		}
	}

	private void AddBallToPlayfield(byte[][] field)
	{
		int x = System.Math.Max(0, System.Math.Min(19, (int)(ballPosition[0])));
		int y = System.Math.Max(0, System.Math.Min(9, (int)(ballPosition[1])));

		field[x][y] = Config.PONG_COLOR_INDEX_BALL;
	}

	private void AddPaddlesToPlayfield(byte[][] field)
	{
		for (int player = 0; player < 2; ++player)
		{
			for (int i = 0; i < PADDLE_SIZE; ++i)
			{
				field[paddlePositionX[player]][paddlePositionY[player] + i] = Config.PONG_COLOR_INDEX_PLAYER[player];
			}
		}
	}

	private void MovePaddle(int player, int change)
	{
		int newPos = paddlePositionY[player] + change;

		if (newPos >= 0 && newPos <= Config.LED_COLS - PADDLE_SIZE)
		{
			paddlePositionY[player] = newPos;
		}
	}

	private static System.Random random = new System.Random();

	private void NewBall()
	{
		ballPosition[0] = (Config.LED_ROWS - 1.0) / 2.0;
		ballPosition[1] = (Config.LED_COLS - 1.0) / 2.0;

		ballDirection[0] = random.NextDouble() > 0.5 ? -0.2 : 0.2;
		ballDirection[1] = random.NextDouble() > 0.5 ? random.NextDouble() * 0.8 + 0.1 : random.NextDouble() * -0.8 - 0.1;
	}

	private void HitLeftRight(int player)
	{
		speaker.BeepSirene();

		NewBall();

		score[player] += 1;

		if (score[player] > 9)
		{
			score[player] = 0;
		}
	}

	private void HitPaddle(double skew)
	{
		speaker.BeepPaddleHit();

		ballDirection[0] = -ballDirection[0];
		ballDirection[1] -= skew;

		for (int i = 0; i < 2; ++i)
		{
			ballDirection[i] *= 1.1; // Increase speed
		}
	}

	private void Tick()
	{
		// Move ball
		for (int i = 0; i < 2; ++i)
		{
			ballPosition[i] += ballDirection[i];
		}

		// Wall collision top/bottom
		if (ballPosition[1] < 0 || ballPosition[1] >= Config.LED_COLS)
		{
			ballDirection[1] = -ballDirection[1];
		}

		// Wall collision left/right
		if (ballPosition[0] < 0)
		{
			HitLeftRight(1);
		}

		if (ballPosition[0] >= Config.LED_ROWS)
		{
			HitLeftRight(0);
		}

		// Paddle collision
		if (ballDirection[0] < 0)
		{
			if (paddlePositionX[0] + 0.5 <= ballPosition[0] && ballPosition[0] <= paddlePositionX[0] + 1.5)
			{
				if (paddlePositionY[0] - 0.5 <= ballPosition[1] && ballPosition[1] <= paddlePositionY[0] + PADDLE_SIZE + 0.5)
				{
					double paddleSkew = (paddlePositionY[0] + PADDLE_SIZE/2.0 - ballPosition[1])/10.0;
					HitPaddle(paddleSkew);
				}
			}
		}

		if (ballDirection[0] > 0)
		{
			if (paddlePositionX[1] - 0.5 <= ballPosition[0] && ballPosition[0] <= paddlePositionX[1] + 0.5)
			{
				if (paddlePositionY[1] - 0.5 <= ballPosition[1] && ballPosition[1] <= paddlePositionY[1] + PADDLE_SIZE + 0.5)
				{
					double paddleSkew = (paddlePositionY[1] + PADDLE_SIZE/2.0 - ballPosition[1])/10.0;
					HitPaddle(paddleSkew);
				}
			}
		}
	}

	public void RunGameLoop()
	{
		FrameRenderedCB(ledStrip, 0);

		timer = new Timer(delegate(object state) { Tick(); }, null, 100, 100);

		while (loop)
		{
			char key = keyPress.ReadSingleKeypress();

			switch (key)
			{
			case 'a':
				MovePaddle(0, -1);
				break;

			case 's':
				MovePaddle(0, 1);
				break;

			case 'k':
				MovePaddle(1, -1);
				break;

			case 'l':
				MovePaddle(1, 1);
				break;

			case 'r':
				InitGame();
				break;

			case 'q':
				loop = false;
				break;
			}
		}

		ledStrip.FrameRendered -= FrameRenderedCB;
		timer.Change(Timeout.Infinite, Timeout.Infinite);
	}

	private static Pong pong;

	static void Main()
	{
		// Create IP Connection and connect it
		IPConnection ipcon = new IPConnection();
		ipcon.Connect(Config.HOST, Config.PORT);

		// Create Tetris object and start game loop
		pong = new Pong(ipcon);

		if (pong.IsOkay()) {
			System.Console.WriteLine("Press q to exit");
			pong.RunGameLoop();
		}

		ipcon.Disconnect();
	}
}

File: games/python/pong.py
#!/usr/bin/env python
# -*- coding: utf-8 -*

from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_led_strip import LEDStrip
from tinkerforge.bricklet_piezo_speaker import PiezoSpeaker

import random
import copy
import time

import config

from threading import Thread

from repeated_timer import RepeatedTimer
from keypress import KeyPress


class PongSpeaker:
    def __init__(self, ipcon):
        self.okay = False
        self.ipcon = ipcon

        if not config.UID_PIEZO_SPEAKER_BRICKLET:
            print("Not Configured: Piezo Speaker")
            return

        self.speaker = PiezoSpeaker(config.UID_PIEZO_SPEAKER_BRICKLET, self.ipcon)

        try:
            self.speaker.get_identity()
            print("Found: Piezo Speaker ({0})".format(config.UID_PIEZO_SPEAKER_BRICKLET))
        except:
            print("Not Found: Piezo Speaker ({0})".format(config.UID_PIEZO_SPEAKER_BRICKLET))
            return

        self.okay = True

    def sirene(self, freq):
        if not self.okay:
            return

        for j in range(2):
            for i in range(25):
                self.speaker.beep(10, freq + i*20)
                time.sleep(0.007)
            for i in range(25):
                self.speaker.beep(10, freq + 24*20 - i*20)
                time.sleep(0.007)

    def beep_paddle_hit(self):
        if not self.okay:
            return

        self.speaker.beep(100, 500)

    def beep_sirene(self):
        if not self.okay:
            return

        Thread(target=self.sirene, args=(1000,)).start()


class Pong:
    PADDLE_SIZE = 3

# Antialased ball?
#    PONG_COLOR_INDEX_BALL_TOP = 8
#    PONG_COLOR_INDEX_BALL_LEFT = 9
#    PONG_COLOR_INDEX_BALL_RIGHT = 10
#    PONG_COLOR_INDEX_BALL_BOTTOM = 11

    COLORS = [
        (  0,   0,   0), # off
#       ( 10,  10,  10), # grey
        (255,   0,   0), # red
        (255,  80,   0), # orange
        (255, 255,   0), # yellow
        (  0, 255,   0), # green
        (  0,   0, 255), # blue
        (255,   0, 150), # violet
        (255,   0,  40), # purple
        (  0,   0,   0), # ball top
        (  0,   0,   0), # ball left
        (  0,   0,   0), # ball right
        (  0,   0,   0)  # ball bottom
    ]

    SCORE_FONT = {
        0: ["222",
            "202",
            "202",
            "202",
            "222"],
        1: ["020",
            "020",
            "020",
            "020",
            "020"],
        2: ["222",
            "002",
            "222",
            "200",
            "222"],
        3: ["222",
            "002",
            "222",
            "002",
            "222"],
        4: ["202",
            "202",
            "222",
            "002",
            "002"],
        5: ["222",
            "200",
            "222",
            "002",
            "222"],
        6: ["222",
            "200",
            "222",
            "202",
            "222"],
        7: ["222",
            "002",
            "002",
            "002",
            "002"],
        8: ["222",
            "202",
            "222",
            "202",
            "222"],
        9: ["222",
            "202",
            "222",
            "002",
            "002"],
    }

    playfield = [x[:] for x in [[0]*config.LED_COLS]*config.LED_ROWS]
    score = [0, 0]
    paddle_position_x = [4, 15]
    paddle_position_y = [3, 3]
    ball_position = [10, 5]
    ball_direction = [0.1, 0.2]
    timer = None
    loop = True

    def __init__(self, ipcon):
        self.okay = False
        self.ipcon = ipcon

        if not config.UID_LED_STRIP_BRICKLET:
            print("Not Configured: LED Strip (required)")
            return

        self.led_strip = LEDStrip(config.UID_LED_STRIP_BRICKLET, self.ipcon)

        try:
            self.led_strip.get_frame_duration()
            print("Found: LED Strip ({0})".format(config.UID_LED_STRIP_BRICKLET))
        except:
            print("Not Found: LED Strip ({0})".format(config.UID_LED_STRIP_BRICKLET))
            return

        self.kp = KeyPress(self.ipcon)
        self.speaker = PongSpeaker(self.ipcon)

        self.okay = True

        self.led_strip.set_frame_duration(40)
        self.led_strip.register_callback(self.led_strip.CALLBACK_FRAME_RENDERED,
                                         self.frame_rendered)

        self.init_game()

    def init_game(self):
        self.new_ball()
        self.paddle_position_y = [3, 3]
        self.score = [0, 0]

    def frame_rendered(self, length):
        self.write_playfield()

    def write_playfield(self):
        if not self.okay:
            return

        field = copy.deepcopy(self.playfield)

        self.add_score_to_playfield(field)
        self.add_paddles_to_playfield(field)
        self.add_ball_to_playfield(field)

        # Reorder LED data into R, G and B channel
        r = []
        g = []
        b = []
        for row in range(config.LED_ROWS):
            col_range = range(config.LED_COLS)
            if row % 2 == 0:
                col_range = reversed(col_range)
            for col in col_range:
                r.append(self.COLORS[field[row][col]][config.R_INDEX])
                g.append(self.COLORS[field[row][col]][config.G_INDEX])
                b.append(self.COLORS[field[row][col]][config.B_INDEX])

        # Make chunks of size 16
        r_chunk = [r[i:i+16] for i in range(0, len(r), 16)]
        g_chunk = [g[i:i+16] for i in range(0, len(g), 16)]
        b_chunk = [b[i:i+16] for i in range(0, len(b), 16)]

        for i in range(len(r_chunk)):
            length = len(r_chunk[i])

            # Fill up chunks with zeros
            r_chunk[i].extend([0]*(16-len(r_chunk[i])))
            g_chunk[i].extend([0]*(16-len(g_chunk[i])))
            b_chunk[i].extend([0]*(16-len(b_chunk[i])))

            try:
                self.led_strip.set_rgb_values(i*16, length, r_chunk[i], g_chunk[i], b_chunk[i])
            except:
                break

    def add_score_to_playfield(self, field):
        for row in range(3):
            for col in range(5):
                field[row][col+1] = int(self.SCORE_FONT[self.score[0]][col][row])
                field[row+17][col+1] = int(self.SCORE_FONT[self.score[1]][col][row])

    def add_ball_to_playfield(self, field):
        x = max(0, min(19, int(self.ball_position[0])))
        y = max(0, min(9, int(self.ball_position[1])))
        field[x][y] = config.PONG_COLOR_INDEX_BALL

# Antialased ball?
#        x = max(0, min(19, self.ball_position[0]))
#        y = max(0, min(9, self.ball_position[1]))
#        ix = int(x)
#        iy = int(y)
#        field[ix][iy] = config.PONG_COLOR_INDEX_BALL
#        if ix + 1 < config.LED_ROWS:
#            field[ix+1][iy] = PONG_COLOR_INDEX_BALL_RIGHT
#        if ix - 1 > 0:
#            field[ix-1][iy] = PONG_COLOR_INDEX_BALL_LEFT
#        if iy + 1 < config.LED_COLS:
#            field[ix][iy+1] = PONG_COLOR_INDEX_BALL_TOP
#        if iy - 1 > 0:
#            field[ix][iy-1] = PONG_COLOR_INDEX_BALL_BOTTOM
#
#        dx = x - int(x)
#        dy = x - int(x)
#        self.COLORS[PONG_COLOR_INDEX_BALL_RIGHT] = (0, 255*dx/64, 0)
#        self.COLORS[PONG_COLOR_INDEX_BALL_LEFT] = (0, 255*(1-dx)/64, 0)
#        self.COLORS[PONG_COLOR_INDEX_BALL_TOP] = (0, 255*dy/64, 0)
#        self.COLORS[PONG_COLOR_INDEX_BALL_BOTTOM] = (0, 255*(1-dy)/64, 0)

    def add_paddles_to_playfield(self, field):
        for player in range(2):
            for i in range(self.PADDLE_SIZE):
                field[self.paddle_position_x[player]][self.paddle_position_y[player]+i] = config.PONG_COLOR_INDEX_PLAYER[player]

    def move_paddle(self, player, change):
        new_pos = self.paddle_position_y[player] + change
        if new_pos >= 0 and new_pos <= config.LED_COLS - self.PADDLE_SIZE:
            self.paddle_position_y[player] = new_pos

    def new_ball(self):
        self.ball_position = [(config.LED_ROWS - 1.0) / 2.0, (config.LED_COLS - 1.0) / 2.0]
        self.ball_direction = [random.choice([-0.2, 0.2]), random.choice([random.randrange(1, 9)/10.0, random.randrange(-9, -1)/10.0])]

    def tick(self):
        # Move ball
        for i in range(2):
            self.ball_position[i] += self.ball_direction[i]

        # Wall collision top/bottom
        if self.ball_position[1] < 0 or self.ball_position[1] >= config.LED_COLS:
            self.ball_direction[1] = -self.ball_direction[1]

        # Wall collision left/right
        def hit_left_right(player):
            self.speaker.beep_sirene()
            self.new_ball()

            self.score[player] += 1
            if self.score[player] > 9:
                self.score[player] = 0

        if self.ball_position[0] < 0:
            hit_left_right(1)

        if self.ball_position[0] >= config.LED_ROWS:
            hit_left_right(0)

        # Paddle collision
        def hit_paddle(skew):
            self.speaker.beep_paddle_hit()
            self.ball_direction[0] = -self.ball_direction[0]
            self.ball_direction[1] -= skew
            for i in range(2):
                self.ball_direction[i] *= 1.1 # Increase speed

        if self.ball_direction[0] < 0:
            if self.paddle_position_x[0] + 0.5 <= self.ball_position[0] <= self.paddle_position_x[0] + 1.5:
                if self.paddle_position_y[0] - 0.5 <= self.ball_position[1] <= self.paddle_position_y[0] + self.PADDLE_SIZE + 0.5:
                    paddle_skew = (self.paddle_position_y[0] + self.PADDLE_SIZE/2.0 - self.ball_position[1])/10.0
                    hit_paddle(paddle_skew)

        if self.ball_direction[0] > 0:
            if self.paddle_position_x[1] - 0.5 <= self.ball_position[0] <= self.paddle_position_x[1] + 0.5:
                if self.paddle_position_y[1] - 0.5 <= self.ball_position[1] <= self.paddle_position_y[1] + self.PADDLE_SIZE + 0.5:
                    paddle_skew = (self.paddle_position_y[1] + self.PADDLE_SIZE/2.0 - self.ball_position[1])/10.0
                    hit_paddle(paddle_skew)

    def run_game_loop(self):
        self.frame_rendered(0)

        self.timer = RepeatedTimer(0.1, self.tick)

        while self.loop:
            key = self.kp.read_single_keypress().lower()

            if key == 'a':
                self.move_paddle(0, -1)
            elif key == 's':
                self.move_paddle(0, 1)
            elif key == 'k':
                self.move_paddle(1, -1)
            elif key == 'l':
                self.move_paddle(1, 1)
            elif key == 'r':
                self.init_game()
            elif key == 'q':
                break

        self.led_strip.register_callback(self.led_strip.CALLBACK_FRAME_RENDERED, None)
        self.timer.stop()


if __name__ == "__main__":
    # Create IP Connection and connect it
    ipcon = IPConnection()
    ipcon.connect(config.HOST, config.PORT)

    # Create Pong object and start game loop
    pong = Pong(ipcon)

    if pong.okay:
        print('Press q to exit')

        pong.run_game_loop()
        pong.kp.kbi.restore_stdin()

    ipcon.disconnect()
